Add `kubectl apply diff-last-applied` subcommand

shiywang opened the pull request:
https://github.com/kubernetes/community/pull/287
`apply diff-last-applied` will invoke `git diff --no-index`, if there's no `git`, it will degenerate to use `diff` also you can use a customize diff tool by setting env : `KUBECTL_EXTERNAL_DIFF`
```console
[root@shiywang template]# ./kubectl apply diff-last-applied -f nginx.yaml 
diff --git a/tmp/kubectl-diff-yd44r.yaml b/tmp/kubectl-diff-0wox0.yaml
index 46aa188..d44cab2 100644
--- a/tmp/kubectl-diff-yd44r.yaml
+++ b/tmp/kubectl-diff-0wox0.yaml
@@ -8,7 +8,7 @@ spec:
   template:
     metadata:
       labels:
-        app: nginx123
+        app: nginx
     spec:
       containers:
       - image: gcr.io/google_containers/nginx-slim:0.8
[root@shiywang template]# 
```

@kubernetes/sig-cli-pr-reviews 
@mengqiy

Commit:
add `kubectl apply diff-last-applied` subcommand

monopole commented:
Hi @shiywang ; 
as part of an effort to eventually move kubectl out of k8s repo, build rules have been installed.
This PR needs something like #46584; you can either /lgtm that, or incorporate it directly here but it may need more approvals.

monopole commented on pkg/kubectl/cmd/BUILD:
```diff
@@ -1,3 +1,5 @@
+package(default_visibility = ["//visibility:public"])
```
No, please don't do this.  

Don't want everything to be able to depend on this package.

shiywang replied:
I think the problem is when I execute `./hack/update-bazel` it will auto generated those files for me, but visibility is wrong, could we auto generate the right one ? so I don't have to manually edit those files.

monopole commented on pkg/kubectl/cmd/BUILD:
```diff
@@ -65,9 +160,6 @@ go_library(
     tags = ["automanaged"],
-    visibility = [
-        "//build/visible_to:pkg_kubectl_cmd_CONSUMERS",
-    ],
```
Please leave this.  

See https://github.com/kubernetes/kubernetes/blob/master/build/visible_to/README.md

monopole commented on pkg/kubectl/cmd/BUILD:
```diff
@@ -265,7 +260,4 @@ filegroup(
     tags = ["automanaged"],
-    visibility = [
-        "//build/visible_to:pkg_kubectl_cmd_CONSUMERS",
-    ],
```
leave this.

pwittrock commented:
/assign @apelisse

pwittrock commented:
/assign apelisse

mengqiy commented:
/assign

mengqiy commented:
@pwittrock Can you confirm if we want to use 2-way diff or 3-way diff in this PR?
I saw 2-way diff in the proposal https://github.com/kubernetes/community/blob/master/contributors/design-proposals/sig-cli/kubectl_apply_getsetdiff_last_applied_config.md
and 3-way diff in https://github.com/kubernetes/kubernetes/pull/44544#pullrequestreview-38471494.

pwittrock commented:
Ideally both controlled by a flag, but we can start with either.  What would the 2-way diff be between?  I assume the live config and the modified config, which would contain fields not present in the config file.  That is fine though.

mengqiy commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -111,53 +125,87 @@ func (e Editor) args(path string) []string {
-		return err
+func (t CmdTool) Launch(paths []string) error {
+	args := make([]string, len(t.Args))
+	copy(args, t.Args)
```
Why copy this?

shiywang replied:
I thought those two are equivalent 
```
args := make([]string, len(t.Args))
copy(args, t.Args)
```
```
args := t.Args
```
just follow the same style [here](https://github.com/kubernetes/kubernetes/pull/46560/files#diff-3266a8985ae8c90ba82083a7fa12f10eR115), CMIIW

mengqiy commented:
> What would the 2-way diff be between?

It seems to be last-applied vs modifed config right now.

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,190 @@
+		if err != nil {
+			return nil
+		}
+		delete(targetObj.(runtime.Unstructured).UnstructuredContent()["metadata"].(map[string]interface{}), "annotations")
```
It is not safe to delete all of the annotations.

shiywang replied:
I'm deleting the annotations in  [line16](https://github.com/kubernetes/kubernetes/blob/4597658cb9bf03640816b561c0ee036c4c1edcd5/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-list/4.response#L16) instead of [line15](https://github.com/kubernetes/kubernetes/blob/4597658cb9bf03640816b561c0ee036c4c1edcd5/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-list/4.response#L15), is that suppose to be always empty like this `\"annotations\":{},\"`?

mengqiy replied:
There maybe other annotations, you cannot delete them all.

shiywang replied:
if we don't remove the annotation, create a new file with `kubectl apply -f rc.yaml`
and then `kubectl apply diff-last-applied -f ~/template/rc.yaml` we expect nothing changed, but actually it will 
be like this
```console
diff --git a/tmp/kubectl-diff-last-applied-original-182tv.yaml b/tmp/kubectl-diff-last-applied-modified-jzjpe.yaml
index 99835c1..11c6832 100644
--- a/tmp/kubectl-diff-last-applied-original-182tv.yaml
+++ b/tmp/kubectl-diff-last-applied-modified-jzjpe.yaml
@@ -1,7 +1,6 @@
 apiVersion: v1
 kind: ReplicationController
 metadata:
-  annotations: {}
   name: nginx
   namespace: default
 spec:
```

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,190 @@
+	switch o {
+	case "json":
+		jsonBuffer := &bytes.Buffer{}
+		err := json.Indent(jsonBuffer, []byte(input), "", "  ")
```
What will happen if `input` is already indented? Will it be double indented?

shiywang replied:
maybe, but I think the input parameter should be raw byte[] instead of indented data, any ideas on this ? comment this function like please do not misuse ? or we do some check in this function to make sure that won't happen ?

shiywang commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -73,34 +81,40 @@ func defaultEnvShell() []string {
+		case "editor":
+			cmd = platformize(defaultEditor, windowsEditor)
+		case "diff":
+			cmd = checkGitDiffInstalled(gitDiff, defaultDiff)
```
sorry, that naming was a little confusing, the defaultDiff means system default tool, not default using it, but I'll change the parameter order, or we could change name, like gitDiff, Diff ?

mengqiy replied:
Please rename them. Or at least leave a comment.

shiywang commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return err
+	}
+	diff := cmdutil.NewDefaultCmdTool("diff", f.DiffEnvs())
```
still one concern here, if user set an uncorrect env, like `KUBECTL_EXTERNAL_DIFF=rm`  it will do some bad things, do we want to limit that kind of operation ? @mengqiy

mengqiy replied:
> KUBECTL_EXTERNAL_DIFF=rm

It may happen. But it may not hurt us. It will remove 2 temp files.

shiywang replied:
ok

alexandercampbell commented:
/assign

alexandercampbell commented on pkg/kubectl/cmd/testdata/diff/test_diff.sh:
```diff
@@ -0,0 +1,29 @@
+if command -v git &>/dev/null; then
+    git diff --no-index "${1}" "${2}" |grep containerPort -A 1 > "/tmp/git-diff.result"
+elif command -v diff &>/dev/null; then
+    diff "${1}" "${2}" |grep containerPort > "/tmp/diff.result"
```
Why write to a unique location? Just write to `./test-result.diff`.

shiywang replied:
updated

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,190 @@
+		o.Output = "json"
+		return nil
+		// If flag -o is not specified, use yaml as default
+	case "yaml", "":
```
Doesn't the command default to `yaml`?

shiywang replied:
seems this function is not used anymore, I'll remove it

pwittrock commented:
@shiywang fyi, we are lifting code freeze this week.  how close is this?  would be good to get merged.

mengqiy commented:
I will pick this up today.

mengqiy commented on pkg/kubectl/cmd/apply_diff_last_applied.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+var (
+	applyDiffLastAppliedLong = templates.LongDesc(`
```
We should support `i18n`, right?

shiywang replied:
updated

mengqiy commented on pkg/kubectl/cmd/apply_diff_last_applied.go:
```diff
@@ -0,0 +1,62 @@
+var (
+	applyDiffLastAppliedLong = templates.LongDesc(`
+		Opens up a 2-way diff in the default diff viewer. This should follow the same semantics as git diff.
+		It should accept either a flag --diff-viewer=meld or check the environment variable KUBECTL_EXTERNAL_DIFF=meld.
```
> KUBECTL_EXTERNAL_DIFF

It becomes `KUBECTL EXTERNAL DIFF` after printing out. But it work for `KUBE_EDITOR` in edit command.
Please investigate. Maybe it is related to `i18n`.

shiywang replied:
the problem is [here](https://github.com/shiywang/kubernetes/blob/2b178ad608bd443efeaf3e9ea7c700e72ab68de1/pkg/kubectl/cmd/templates/normalizers.go#L73), this markdown function and `ASCIIRenderer` @fabianofranz do you have any ideas on this ? 
and more accurately `KUBE_EDITOR` also have problem, it render `KUBE_EDITOR` to `KUBE_ EDITOR` if you watch carefully

mengqiy replied:
@shiywang Do you have an idea how to fix this?

shiywang replied:
how about file an issues and assign to me, I will fix that in a separated pr since I'm not sure how long it will take.

mengqiy commented:
```
$ kubectl apply diff-last-applied -f ../depl.yaml
diff --git a/tmp/kubectl-diff-nyxxv.yaml b/tmp/kubectl-diff-i7hvs.yaml
index 1b30d2591d..2bd15b2a20 100644
--- a/tmp/kubectl-diff-nyxxv.yaml
+++ b/tmp/kubectl-diff-i7hvs.yaml
@@ -1,8 +1,6 @@
 apiVersion: extensions/v1beta1
 kind: Deployment
 metadata:
-  annotations:
-    hello: world
   labels:
     app: foo
```
When I add something in the config file, it should use `+` instead of `-`.  Otherwise, it is counter-intuitive.

mengqiy commented:
I think we should use more meaningful file names, like `kubectl-apply-diff-last-applied-original-xxxxx` and `kubectl-apply-diff-last-applied-modified-xxxxx`.
We should remove these 2 temp files in the end.

And it will also be helpful to print out what diff tool it is using on each run.

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+//TODO: same function in apply_view_last_applied.go, find a better way to reuse ?
+func (o *DiffOptions) ValidateOutputArgs(cmd *cobra.Command) error {
```
It seems this function is not used elsewhere. So we can remove it.

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+func (o *DiffOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
+	if o.Output != "" {
```
The `-o` flag is defaulted to `yaml`. Anything that is neither "yaml" nor "json" should return an error, including "".

shiywang commented:
@mengqiy updated, PTAL

shiywang commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+		}
+		return yamlOutput, nil
+	}
+	return nil, fmt.Errorf("unexpected -o output mode")
```
in [style convention](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) this should not be capitalized, but I saw some command's error message is like this `error: Unexpected .... `  is this mean if the error message is at the beginning of `error:` we should capitalized ? do we have convention of this ? @pwittrock

alexandercampbell replied:
When I doubt I always follow the style guide.
Following the style guide gives a few advantages:

1. Other people will be more likely to re-use our code since it matches their expectations.
2. New developers who are used to the typical style can contribute without additional context.
3. Go tools like "golint" work without any changes.

shiywang commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -61,20 +61,31 @@ func validateApplyArgs(cmd *cobra.Command, args []string) error {
+	if err != nil && os.IsNotExist(err) {
+		return false
+	}
+	return true
```
updated

shiywang commented:
/retest

alexandercampbell commented on pkg/kubectl/cmd/apply_diff_last_applied.go:
```diff
@@ -0,0 +1,63 @@
+var (
+	applyDiffLastAppliedLong = templates.LongDesc(i18n.T(`
+		Opens up a 2-way diff in the default diff viewer. This should follow the same semantics as git diff.
+		It should accept either a flag --diff-viewer=meld or check the environment variable KUBECTL_EXTERNAL_DIFF=meld.
```
I like how `KUBECTL_EXTERNAL_DIFF` matches `GIT_EXTERNAL_DIFF`.

shiywang replied:
@mengqiy @pwittrock wdyt ?

alexandercampbell replied:
(if my comment was unclear: I meant I liked this variable name. Good choice!)

mengqiy replied:
Maybe we drop `CTL`, using `KUBE_EXTERNAL_DIFF` like `KUBE_EDITOR`.

alexandercampbell replied:
If there's already a `KUBE_EDITOR`, we should definitely use `KUBE_EXTERNAL_DIFF`.

alexandercampbell commented on pkg/kubectl/cmd/apply_diff_last_applied.go:
```diff
@@ -0,0 +1,63 @@
+		},
+	}
+
+	cmd.Flags().StringVarP(&options.Output, "output", "o", "yaml", "Output format. One of: yaml|json.")
```
There should be a utility method to define this flag in cmdutil.

shiywang replied:
same above, @mengqiy @pwittrock wdyt ? I'm ok with that but just curious when should we move flags into helpers ? this flag only invoked at most twice, is it worth to move into helpers ?

mengqiy replied:
+1 for having a helper function to add this flag.
I thought we already have one for `--output`. It will surprise me if we don't.

alexandercampbell replied:
>at most twice

Even two times would be enough for this to be consolidated to a single helper function. Avoid duplicating code and especially avoid duplicating string literals.

Incidentally, `--output` is exceedingly common and is defined at least 7 times in kubectl.

```
$ rg "Flags" | rg "String" | rg '"output"' | rg -v _test 
1:6:25:cmd/apply_edit_last_applied.go:98:	cmd.Flags().StringVarP(&options.Output, "output", "o", "yaml", "Output format. One of: yaml|json.")
2:7:27:cmd/apply_view_last_applied.go:75:	cmd.Flags().StringP("output", "o", "", "Output format. Must be one of yaml|json")
3:17:97:cmd/edit.go:109:	cmd.Flags().StringVarP(&options.Output, "output", "o", "yaml", "Output format. One of: yaml|json.")
4:55:262:cmd/version.go:59:	cmd.Flags().String("output", "", "output format, options available are yaml and json")
5:144:788:cmd/util/printing.go:56:	cmd.Flags().StringP("output", "o", "", "Output mode. Use \"-o name\" for shorter output (resource/name).")
6:145:791:cmd/util/printing.go:61:	cmd.Flags().StringVarP(output, "output", "o", "", "Output mode. Use \"-o name\" for shorter output (resource/name).")
7:146:794:cmd/util/printing.go:66:	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|wide|name|custom-columns=...|custom-columns-file=...|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See custom columns [http://kubernetes.io/docs/user-guide/kubectl-overview/#custom-columns], golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://kubernetes.io/docs/user-guide/jsonpath].")
```

mengqiy replied:
We should clean up this.

shiywang replied:
@mengqiy @alexandercampbell  done, ptal

alexandercampbell commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -61,20 +61,28 @@ func validateApplyArgs(cmd *cobra.Command, args []string) error {
+func pathExist(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil || !os.IsNotExist(err)
+}
```
Use `kubernetes/util.FileExists` instead of writing this yourself.

shiywang replied:
ok, will do that

alexandercampbell commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -990,6 +998,91 @@ func TestRunApplySetLastApplied(t *testing.T) {
+		}
+		tf.Namespace = "test"
+		tf.ClientConfig = defaultClientConfig()
+		buf, errBuf := bytes.NewBuffer([]byte{}), bytes.NewBuffer([]byte{})
```
```go
buf, errBuf := new(bytes.Buffer), new(bytes.Buffer)
```

shiywang replied:
what's the difference ?

alexandercampbell replied:
from the [Go documentation](https://golang.org/pkg/bytes/#NewBuffer):
> In most cases, new(Buffer) (or just declaring a Buffer variable) is sufficient to initialize a Buffer.

There is no behavior difference. I have a personal preference for the most simple or common pattern, which in this case is new(Buffer). Your call.

alexandercampbell commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -990,6 +998,91 @@ func TestRunApplySetLastApplied(t *testing.T) {
+		if pathExist("/tmp/git-diff.result") {
+			out := string(readBytesFromFile(t, "/tmp/git-diff.result"))
+			if out != test.expectedGitDiffOut {
+				t.Fatalf("%s: unexpected output: %s\nexpected: %s", test.name, out, test.expectedGitDiffOut)
```
May be worth using %q here to more clearly identify whitespace issues.

shiywang replied:
sounds reasonable

alexandercampbell commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -990,6 +998,91 @@ func TestRunApplySetLastApplied(t *testing.T) {
+		cmd.Flags().Set("output", test.output)
+		cmd.Run(cmd, []string{})
+
+		if pathExist("/tmp/git-diff.result") {
```
This is broken. The test tests nothing when `/tmp/git-diff.result` does not exist. Rewrite the test to check the file regardless of whether it exists.

shiywang replied:
ah, my mistake, will update soon

alexandercampbell replied:
Thanks!

alexandercampbell commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -990,6 +998,91 @@ func TestRunApplySetLastApplied(t *testing.T) {
+			os.Remove("/tmp/git-diff.result")
+		}
+
+		if pathExist("/tmp/diff.result") {
```
These filenames definitely need to be constants defined at the top of the function.

shiywang replied:
ok

alexandercampbell commented on pkg/kubectl/cmd/testdata/diff/test_diff.sh:
```diff
@@ -0,0 +1,29 @@
+    git diff --no-index "${1}" "${2}" |grep containerPort -A 1 > "./git-diff.result"
+elif command -v diff &>/dev/null; then
+    diff "${1}" "${2}" |grep containerPort > "./diff.result"
+else
```
You need both of these to happen, no?

Don't add this complicated conditional logic. Just do

```sh
git diff --no-index "${1}" "${2}" |grep containerPort -A 1 > "./git-diff.result"
diff "${1}" "${2}" |grep containerPort > "./diff.result"
```

The script will fail if those programs are not installed.

shiywang replied:
I don't need both of these happen, if `git diff` exist it will use `git diff` and omit `diff`, I also update unit tests to make sure as long as one of those pass tests will pass, and this script is similar to `test_editor.sh`, I think this is fine.

alexandercampbell replied:
We should pick just one. `git diff` and `diff` have slightly different behaviors. Reasoning: the unit tests should never behave differently on the basis of the local development environment.

`git diff` makes more sense to me just because the user needs to have `git` to clone the repository anyway.

shiywang replied:
@mengqiy wdyt ?

alexandercampbell reviewed (changes requested):
I've found a few issues so far. Let me know what you think of my comments and I will come back to finish the review.

shiywang commented:
@mengqiy  @alexandercampbell  PTAL

mengqiy commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -73,34 +81,40 @@ func defaultEnvShell() []string {
 
-func defaultEnvEditor(envs []string) ([]string, bool) {
-	var editor string
+func defaultEnv(cmdName string, envs []string) ([]string, bool) {
```
> cmdName string

I suggest you define a type for this to avoid random string get passed in.
```
type CmdName string
const(
  editorCmd CmdName = "editor"
  diffCmd CmdName = "diff"
)
```

mengqiy commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -73,34 +81,40 @@ func defaultEnvShell() []string {
-	args := make([]string, len(e.Args))
-	copy(args, e.Args)
-	if e.Shell {
+func (t CmdTool) args(this []string, path string) []string {
```
Why passing an seperate variable `this`?
Why not just use `t.Args`?

shiywang replied:
updated, how about now ?

mengqiy commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -190,3 +235,12 @@ func platformize(linux, windows string) string {
 }
+
+func checkGitDiffInstalled(gitdiff, diff string) string {
+	cmd := exec.Command("git", "version")
```
Is using `hash git` make for sense here? But the question is does `hash` works on windows?

mengqiy replied:
Or use `command git`?

mengqiy commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -61,18 +62,24 @@ func validateApplyArgs(cmd *cobra.Command, args []string) error {
 	filenameWidgetClientside = "../../../test/fixtures/pkg/kubectl/cmd/apply/widget-clientside.yaml"
 	filenameWidgetServerside = "../../../test/fixtures/pkg/kubectl/cmd/apply/widget-serverside.yaml"
+
+	fileNameGitDiffResult    = "./git-diff.result"
```
I will save temp file in "/tmp". This will avoid giving you a dirty tree when tests are failing or interrupted.

shiywang replied:
yes, previously I was saving temp file in /tmp, but @alexandercampbell suggest to save in `./` which I'm confuse now... : -/

shiywang replied:
I prefer saving in `/tmp` if you guys have no objection, I'll change it back

alexandercampbell replied:
@shiywang I had asked you to use "./" to avoid relying on the presence of a "/tmp" directory. It's a minor point but it could break the tests on an unusual platform or in a situation with unusual permissions. I like @mengqiy's suggestion, but instead of "/tmp", try https://golang.org/pkg/io/ioutil/#TempFile. I think that would be the best of both worlds.

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+	switch o {
+	case "json":
+		jsonBuffer := &bytes.Buffer{}
+		err := json.Indent(jsonBuffer, []byte(input), "", "  ")
```
No need to do type assertion for `input`.

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+		}
+		return jsonBuffer.Bytes(), nil
+	case "yaml":
+		yamlOutput, err := yaml.JSONToYAML([]byte(input))
```
No need to do type assertion for `input`.

shiywang commented:
/retest

shiywang commented:
/test pull-kubernetes-cross

alexandercampbell commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -990,6 +997,97 @@ func TestRunApplySetLastApplied(t *testing.T) {
+			continue
+		}
+
+		t.Fatalf("both diff tools not exist or file operation error")
```
How about this?
```go
t.Fatalf(`External test script failed. Check that either "diff" or "git diff" is installed.`)
```

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,255 @@
+	// rather than parse the shell arguments ourselves, punt to the shell
+	shell := defaultEnvShell()
+	return append(shell, cmd), true
+}
```
I do *not* like introducing this level of complexity. What's the purpose of this file? Is there no existing utility we can use?

shiywang replied:
sorry I don't understand, could you tell me more specific ?

alexandercampbell replied:
This entire file is extremely complex. There's only a small amount of test code, so we don't know if it works. Even if we did test it, it would be very hard to catch all of the little edge cases (such as whitespace or escaped quotes in filenames or executable names). Is there another way?

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,255 @@
+	dir := os.TempDir()
+
+	for i := 0; i < 10000; i++ {
+		name := filepath.Join(dir, prefix+randSeq(5)+suffix)
```
There are many antipatterns in this function. The biggest one is that you're rerolling a standard library function, [ioutil.TempFile](https://golang.org/pkg/io/ioutil/#TempFile). Whenever you're doing a low-level operation like creating a temp file or directory, just do a quick search on the standard library and you'll probably find what you're looking for.

apelisse replied:
yeah, the potential race condition creates a security risk.

smarterclayton replied:
Chesterton's Fence applies in this case.  The stdlib does not support extensions for temp files, which means temporary files on platforms with extensions do not perform properly.  This is the accepted way to work around that limitation in the stdlib.

https://github.com/openshift/origin/pull/1795

smarterclayton replied:
The method deserves a comment to that effect.

apelisse replied:
ok yeah, I've re-read the code, it's actually fine.

shiywang replied:
ok, will add some comments

alexandercampbell replied:
@smarterclayton I see. Thanks for clearing this up.

@shiywang in that docstring, would you mind including a link to that openshift/origin PR? Duplicated code should at least be aware of its original location.

shiywang replied:
@alexandercampbell ok

alexandercampbell reviewed (changes requested):
I have a few more comments-- will continue the review in a bit. Thanks!

alexandercampbell commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -61,6 +61,16 @@ func AddOutputVarFlagsForMutation(cmd *cobra.Command, output *string) {
 
+// AddOutputFlagsForJsonYaml adds output related flags to a command. But only support json or yaml format.
+func AddOutputFlagsForJsonYaml(cmd *cobra.Command, defaultOutput string) {
+	cmd.Flags().StringP("output", "o", defaultOutput, "Output format. One of: yaml|json.")
```
Just call the other function.

```go
AddOutputVarFlagsForJsonYaml(cmd, new(string), defaultOutput)
```

shiywang replied:
I think this is fine, see AddOutputVarFlagsForMutation

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	windowsEditor = "notepad"
+	windowsShell  = "cmd"
+
+	//for diff command
```
This comment might as well be removed.

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+)
+
+const (
+	// sorry, blame Git
```
What does "sorry, blame Git" mean?

shiywang replied:
sorry, I really don't know, I think we can use `git blame` to find out who wrote this code and cc he/she : )

alexandercampbell replied:
You don't know? Who wrote the code then?

mengqiy replied:
This piece actually comes from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/editor/editor.go#L36.
I will treat this file as renaming then editing. It seems that the review tool doesn't recognise this.

mengqiy replied:
I think we can remove this line, if it doesn't make sense.

alexandercampbell replied:
Hm, interesting. Thanks for pointing that out @mengqiy

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	// sorry, blame Git
+	// TODO: on Windows rely on 'start' to launch the editor associated
+	// with the given file type. If we can't because of the need of
+	// blocking, use a script with 'ftype' and 'assoc' to detect it.
```
What is "the need of blocking"?

shiywang replied:
ditto

alexandercampbell commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -61,6 +61,16 @@ func AddOutputVarFlagsForMutation(cmd *cobra.Command, output *string) {
+}
+
+// AddOutputVarFlagsForJsonYaml adds output related flags to a command. But only support json or yaml format.
+func AddOutputVarFlagsForJsonYaml(cmd *cobra.Command, output *string, defaultOutput string) {
```
I would name this `AddOutputVarFlag` and the other function `AddOutputFlag`.

shiywang replied:
I think this is fine, see `AddOutputVarFlagsForMutation`

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+	tmpFileSlice := []cmdutil.TempFile{
+		//be careful with the order of DiffBufferModified and DiffBufferOriginal
+		{Prefix: fmt.Sprintf("%s-diff-last-applied-original-", filepath.Base(os.Args[0])), Suffix: o.ext, Buffer: bytes.NewReader(o.DiffBufferOriginal)},
+		{Prefix: fmt.Sprintf("%s-diff-last-applied-modified-", filepath.Base(os.Args[0])), Suffix: o.ext, Buffer: bytes.NewReader(o.DiffBufferModified)},
```
Why does this need to be based on `os.Args[0]`? Seems to reduce predictability. What about just hardcoding `"kubectl-diff-last-applied-modified-"`?

shiywang replied:
maybe some people rename their binary name, I don't know, I think we can change it here if we confirm with others.

alexandercampbell replied:
I know people can rename their binaries.
I would prefer that this was hardcoded to "kubectl" for that case. It introduces more entropy to move the files on the basis of the binary name.

shiywang replied:
I'm not gonna change that, because openshift/origin could reuse those code, and hard code will cause problem, see similar issue here: https://github.com/kubernetes/kubernetes/pull/48750

mengqiy replied:
Or maybe just "diff-last-applied-modified-"?

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+		return err
+	}
+
+	return nil
```
This check + return can just be `return err`

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+
+func (o *DiffOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
+	if o.Output != "yaml" && o.Output != "json" {
+		return fmt.Errorf("invalid output format %s, only yaml|json supported", o.Output)
```
I would change
`%s --> %q`

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+		return err
+	}
+	o.Codec = runtime.NewCodec(f.JSONEncoder(), f.Decoder(true))
+	o.ext = "." + o.Output
```
I would be surprised if I noticed `ext` includes the separating period.

shiywang replied:
I think this is ok

mengqiy replied:
I think maybe we can even get rid of `o.ext`, since it doesn't help much. It is just `"." + o.Output`. `o.Output` has been validated, it should not have any issue.

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		return err
+	}
+	return nil
```
Can just `return err`

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+	switch o {
+	case "json":
+		jsonBuffer := &bytes.Buffer{}
+		err := json.Indent(jsonBuffer, input, "", "  ")
```
Shouldn't this read the pretty-print command-line flags?

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+				return err
+			} else {
+				return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%v\nfrom server for:", info), info.Source, err)
+			}
```
```go
if b {
    return err1
} else {
    return err2
}
```
can just be
```go
if b {
    return err1
}
return err2
```

shiywang replied:
updated

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+			return err
+		}
+
+		return nil
```
Can just `return err`

shiywang replied:
yes, you are right

alexandercampbell commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,177 @@
+		if err != nil {
+			return nil, err
+		}
+		return yamlOutput, nil
```
This whole case can be

```go
return yaml.JSONToYAML(input)
```

shiywang replied:
ok will do that

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	return string(b)
+}
+
+func platformize(linux, windows string) string {
```
Maybe rather that `linux` this should be `unix`? `kubectl` can be run on MacOS, no?

shiywang replied:
cc @kargakis

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	return
+}
+
+var letters = []rune("abcdefghijklmnopqrstuvwxyz0123456789")
```
Why this is `rune` instead of `byte`?

shiywang replied:
I think maybe `rune` is also ok, what's your concern here ?

alexandercampbell replied:
`rune` works as well. I'm just curious why it was chosen instead of `byte`.

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	return bytes, paths[0], err
+}
+
+func prepareTempFile(prefix, suffix string, r io.Reader) (string, error) {
```
This function could use a docstring.

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	if !strings.Contains(cmd, " ") {
+		return []string{cmd}, false
+	}
+	if !strings.ContainsAny(cmd, "\"'\\") {
```
I'd use backquotes here so this is more legible.

```go
ContainsAny(cmd, `"'\`)
```

shiywang replied:
ok, will do that

alexandercampbell commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,257 @@
+	// with the given file type. If we can't because of the need of
+	// blocking, use a script with 'ftype' and 'assoc' to detect it.
+	defaultEditor = "vi"
+	defaultShell  = "/bin/bash"
```
Wouldn't it make more sense for these to be called `unixEditor` and `unixShell`?

shiywang replied:
I think this is fine, I don't change the name, I just bump the original code here

alexandercampbell reviewed (changes requested):
Made another pass. Thanks!

shiywang commented:
/test pull-kubernetes-kubemark-e2e-gce

shiywang commented:
@alexandercampbell @mengqiy really appreciate your careful review, could you have a consensus on those unaddressed comments, then I will update it.

pwittrock commented:
/unassign

pwittrock commented:
re-assign to me when this is ready for approval

mengqiy commented:
@shiywang I don't think we have any disagreement. Where?

shiywang commented:
@mengqiy @alexandercampbell update, ptal, most of remain unaddressed comments I just bump from original code which I don't think should solved in this pr, @alexandercampbell feel free to open an new pr to address those if you think it's really bothering you, thanks for you guys patient and sorry for my delay.

mengqiy commented:
/retest
Will do another pass today. Hopefully it's final pass.

mengqiy commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,255 @@
+// and file prefix, and then invokes Launch with the path of that file. It will return
+// the contents of the first file after launch, any errors that occur, and the path of the
+// temporary file so the caller can clean it up as needed.
+func (t CmdTool) LaunchTempFile(tmpFile ...TempFile) ([]byte, string, error) {
```
I suggest you break this function into 2:
```
func (t CmdTool) LaunchTempFiles(tmpFiles ...TempFile) ([]string, error) {
	var paths []string
	for _, f := range tmpFile {
		path, err := prepareTempFile(f.Prefix, f.Suffix, f.Buffer)
		if err != nil {
			return nil, path, err
		}
		paths = append(paths, path)
	}
	return paths, t.Launch(paths)
}
```
```
func (t CmdTool) LaunchTempFile(tmpFile TempFile) ([]byte, string, error) {
    paths, err := LaunchTempFiles(tmpFile)
    if err != nil {
      reutrn nil, "", err
    }
	bytes, err := ioutil.ReadFile(paths[0])
	return bytes, paths[0], err
}
```

mengqiy commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -0,0 +1,255 @@
+	if runtime.GOOS == "windows" {
+		cmdName = "where"
+	}
+	cmd := exec.Command(cmdName, "git")
```
I doubt that if this will work on windows, since git on windows should be "git.exe"?

shiywang replied:
I tried on windows, `where git.exe`  or `where git`  both ok

mengqiy replied:
Good!

mengqiy commented on pkg/kubectl/cmd/apply_set_last_applied.go:
```diff
@@ -200,46 +197,13 @@ func (o *SetLastAppliedOptions) RunSetLastApplied(f cmdutil.Factory, cmd *cobra.
 			if err != nil {
 				return err
 			}
+			fmt.Fprintf(o.Out, string(out))
```
Why we start to print it? Is this for debug?

shiywang replied:
no, this value should pass to PrintSuccess, if we execute `./kubectl apply set-last-applied -f ~/nginx.yaml  -o yaml` it will print out

mengqiy replied:
I finally understood. `FormatConvert` no longer does the printing.

mengqiy commented on pkg/kubectl/cmd/apply_test.go:
```diff
@@ -990,6 +997,97 @@ func TestRunApplySetLastApplied(t *testing.T) {
 	cmdutil.BehaviorOnFatal(func(str string, code int) {})
 }
 
+func TestRunApplyDiffLastApplied(t *testing.T) {
```
I think it will make more sense to move `diff_last_applied`'s test to its own file like what we do for create subcommands tests. 
I'm fine to resolve it in a followup PR. `set_last_applied` is already in this file. And you will also move these to `apply`'s own pkg.

shiywang replied:
yes, I already have a issues to track that, will do that after this pr merged

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,163 @@
+		if err != nil {
+			return nil
+		}
+		delete(targetObj.(runtime.Unstructured).UnstructuredContent()["metadata"].(map[string]interface{}), "annotations")
```
Followup https://github.com/kubernetes/kubernetes/pull/46560#discussion_r119524943.
Once again you can not delete all the annotations. The right thing to do here is deleting annotations only when it is all empty.

You can try the following config:
```
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  annotations:
    hello: world
  name: nginx-deployment
spec:
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
        ports:
        - containerPort: 80
```
With your implementation, it shows the following even there is no change
```
3a4,5
>   annotations:
>     hello: world
```

shiywang replied:
oh, thanks, sorry, I just forgot update here...

shiywang commented:
@mengqiy comments addressed, ptal

mengqiy commented:
/retest

shiywang commented:
/test pull-kubernetes-cross
/test pull-kubernetes-federation-e2e-gce

cblecker commented:
cross-build is currently broken: https://github.com/kubernetes/kubernetes/issues/48887

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -121,6 +123,18 @@ func (o *DiffOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
 	return err
 }
 
+func isAnnotationsEmpty(obj runtime.Object) bool {
```
You don't need to do it by youself.
We already have some handy tools.
See: https://gist.github.com/mengqiy/61189aeecf13660430471aac0c45a978

shiywang commented on pkg/kubectl/cmd/util/cmdtools.go:
```diff
@@ -170,23 +170,31 @@ type TempFile struct {
 		path, err := prepareTempFile(f.Prefix, f.Suffix, f.Buffer)
 		if err != nil {
-			return nil, path, err
+			return []string{""}, err
```
I just want to consist with return value of `prepareTempFile`, but I think returnning `nil` is also fine, I will update it to return `nil`

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -121,6 +128,11 @@ func (o *DiffOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
 }
 
+func deleteEmptyAnnotations(obj runtime.Object) {
+	delete(obj.(runtime.Unstructured).UnstructuredContent()["metadata"].(map[string]interface{}), "annotations")
```
Type assertion without error checking is not completely safe. Use `SetAnnotation()`.
Still see:  https://gist.github.com/mengqiy/61189aeecf13660430471aac0c45a978

shiywang replied:
yes, but I think `SetAnnotation` doesn't work

shiywang replied:
I made a little change, but do you mean directly use `SetAnnotation()` ? that looks good, but unfortunately, it doesn't works as you expected, still not remove annotation field, just set to empty, you can try it yourself.

mengqiy replied:
You are right. It doesn't work as expected. My intend is just to reuse some code here.

I suspect it may be a bug with `SetAnnotations()`, since it cannot distinguish `SetAnnotations(targetObj, nil)` with `SetAnnotations(targetObj, {})`.
Can you please file an issue for it?

shiywang replied:
file a new issue here: https://github.com/kubernetes/kubernetes/issues/49075

dims commented:
/test pull-kubernetes-cross

shiywang commented:
/test pull-kubernetes-cross

dims commented:
@shiywang Can you try rebasing to latest master?

mengqiy commented on pkg/kubectl/cmd/apply_diff_last_applied.go:
```diff
@@ -0,0 +1,62 @@
+var (
+	applyDiffLastAppliedLong = templates.LongDesc(i18n.T(`
+		Opens up a 2-way diff in the default diff viewer. This should follow the same semantics as git diff.
+		It should accept either a flag --diff-viewer=meld or check the environment variable KUBECTL_EXTERNAL_DIFF=meld.
```
> --diff-viewer

You haven't implement this flag yet.

shiywang replied:
I just updated the help message, ptal, do you think this is ok ?

mengqiy commented on pkg/kubectl/cmd/util/diff/diffoptions.go:
```diff
@@ -0,0 +1,183 @@
+
+type DiffOptions struct {
+	FilenameOptions    resource.FilenameOptions
+	Mapper             meta.RESTMapper
```
I found the scope of Mapper, Typer and Codec is limited to `Complete()`.
If that is the case, it is not necessary to put them in the `DiffOptions`.
We should keep it only having what is necessary.

shiywang replied:
done

mengqiy commented:
@shiywang Please fix `hack/verify-golint.sh`.

shiywang commented:
close this due to conversation is too long, open a new pr here

mengqiy commented:
> close this due to conversation is too long, open a new pr here #49174

A lot of discussion is valuable. Actually no need to close.
